File: leetcode-js/1448.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var goodNodes = function (root) {
    if (!root) return 0;
    let max;
    let answer = 0;

    const traverseTree = (node) => {
        if (!node) return;
        if (stack.length === 0) {
            max = node.val;
            answer++;
        }
        else if (max <= node.val) {
            max = node.val;
            answer++;
        }
        let localMax = max;
        traverseTree(node.left);
        traverseTree(node.right);
        max = localMax;
    }
    traverseTree(root);
    return answer;
};

File: leetcode-go/1448.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func traverseFunc(root *TreeNode, maxStack []int) int {
	if root == nil {
		return 0
	}
	result := 0
	if len(maxStack) == 0 {
		result++
		maxStack = append(maxStack, root.Val)
	} else if root.Val > maxStack[len(maxStack)-1] {
		result++
		maxStack = append(maxStack, root.Val)
	}
	result += traverseFunc(root.Left, maxStack)
	result += traverseFunc(root.Right, maxStack)
	if root.Val == maxStack[len(maxStack)-1] {
		maxStack = maxStack[:len(maxStack)-1]
	}
	return result
}

func goodNodes(root *TreeNode) int {
	maxStack := []int{}
	return traverseFunc(root, maxStack)
}
